linux/GPU.c

```diff
@@ -0,0 +1,272 @@
+                  .id = client_id,
+                  .pdev = pdev,
+                  .next = parsed_ids,
+               };
+               pdev = NULL;
+
+               parsed_ids = new;
+            }
+
+            sstate = SECST_UNKNOWN;
+
+            errno = 0;
+            client_id = strtoull(buf + strlen("drm-client-id:"), &endptr, 10);
+            if (errno || *endptr != '\n')
+               client_id = INVALID_CLIENT_ID;
+         } else if (String_startsWith(buf, "drm-pdev:")) {
+            const char* p;
+            char* q;
+
+            p = buf + strlen("drm-pdev:");
+
+            while (isspace((unsigned char)*p))
+               p++;
+
+            q = strchr(p, '\n');
+            if (q)
+               *q = '\0';
+
+            assert(!pdev || String_eq(pdev, p));
+            if (!pdev)
+               pdev = xStrdup(p);
+         } else if (String_startsWith(buf, "drm-engine-")) {
+            const char* delim;
+            const char* engineStart;
+            char* endptr;
+            unsigned long long int value;
+
+            if (sstate == SECST_DUPLICATE)
+               continue;
+
+            if (String_startsWith(buf, "drm-engine-capacity-"))
+               continue;
+
+            delim = strchr(buf, ':');
+            engineStart = buf + strlen("drm-engine-");
+
+            errno = 0;
+            value = strtoull(delim + 1, &endptr, 10);
+            if (errno == 0 && String_startsWith(endptr, " ns")) {
+               if (sstate == SECST_UNKNOWN) {
+                  if (client_id != INVALID_CLIENT_ID && !is_duplicate_client(parsed_ids, client_id, pdev))
+                     sstate = SECST_NEW;
+                  else
+                     sstate = SECST_DUPLICATE;
+               }
+
+               if (sstate == SECST_NEW) {
+                  new_gpu_time += value;
+                  update_system(lpl, value, xStrndup(engineStart, delim - engineStart));
+               }
+            }
+         }
+      } /* finished parsing lines */
+
+      fclose(fp);
+
+      if (sstate == SECST_NEW) {
+         struct client_id* new;
+
+         assert(client_id != INVALID_CLIENT_ID);
+
+         new = xMalloc(sizeof(*new));
+         *new = (struct client_id) {
+            .id = client_id,
+            .pdev = pdev,
+            .next = parsed_ids,
+         };
+         pdev = NULL;
+
+         parsed_ids = new;
+      }
+
+      free(pdev);
+   } /* finished parsing fdinfo entries */
+
+   if (new_gpu_time > 0) {
+      const Machine* host = lp->super.host;
+      unsigned long long int gputimeDelta;
+      uint64_t monotonictimeDelta;
+
+      Process_updateFieldWidth(GPU_TIME, ceil(log10(new_gpu_time)));
+
+      gputimeDelta = saturatingSub(new_gpu_time, lp->gpu_time);
+      monotonictimeDelta = host->monotonicMs - host->prevMonotonicMs;
+      if (gputimeDelta == 0)
+         lp->gpu_percent = 0.0f;
+      else if (monotonictimeDelta == 0)
+         lp->gpu_percent = -1.0f;
```

Comment:
Perhaps I should mention issue #1274 with this part of code.

Reply:
@cgzones monotonictimeDelta can never be zero here, so the 'else if' branch is dead code - please remove.

Reply:
@natoscott Place an assert statement there so that monotonictimeDelta can never be zero.

Reply:
The nature of time and space is what makes the value never zero, not an assert.  :)   But seriously, I'd not object to an assert if it helps others feel at ease, even though personally I think its just noise in the code.

Reminds me of the discussion a few years back about asserting on a non-NULL pointer right before we use it ... which some people like to do, but again personally I feel its just noise in the code since both debug and non-debug builds will inform you immediately and clearly if that ever happens (hence the assert adds no value IMO).

Reply:
@natoscott The assert statement is to prevent bugs. 🙂
Even the variable is named `monotonicTime`, programmers can still retrieve the same cached value twice (usually done accidentally). There is no guarantee that the value would always be different unless they call `clock_gettime` or similar API every time.

Reply:
It doesn't prevent anything - the bug would still happen and if there's no assert, we divide by zero (in both debug *and* non-debug builds mind you) and stop on the exact same line of code.  As with the non-null-pointer-assert, its just line noise in the code IMO.

But as I said, I understand it makes some people feel better - so please go right ahead with an assert if you feel its useful.

Reply:
You're also kinda pointing out the problem with #1274 too - it would be hiding such bugs.  If they happen, we really want to know about them, not just carry on execution as if nothing happened.

Reply:
added an assert in ./Machine.c